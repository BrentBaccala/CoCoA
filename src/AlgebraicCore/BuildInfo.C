//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
