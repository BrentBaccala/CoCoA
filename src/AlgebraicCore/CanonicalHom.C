//   Copyright (c)  2007,2009  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/CanonicalHom.H"
#include "CoCoA/RingHom.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/PolyRing.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/FractionField.H"

// #include <iostream>  // for debugging only

namespace CoCoA
{

  RingHom CanonicalHom(const ring& domain, const ring& codomain)
  {
    if (domain == codomain) return IdentityHom(domain);

    // Two easy cases:
    if (IsZZ(domain)) return ZZEmbeddingHom(codomain);
    if (IsQQ(domain)) return QQEmbeddingHom(codomain); // NB result is only a partial hom!!

    // Check codomain first, as this makes it possible to exploit certain "shortcuts"
    if (IsFractionField(codomain))
    {
      if (domain == BaseRing(codomain))
        return EmbeddingHom(codomain);
      else
	return EmbeddingHom(codomain)(CanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      if (domain == CoeffRing(codomain))
        return CoeffEmbeddingHom(codomain);
      else
	return CoeffEmbeddingHom(codomain)(CanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      const QuotientRing QR = codomain;
      if (domain == BaseRing(QR))
        return QuotientingHom(QR);
      else
	return QuotientingHom(QR)(CanonicalHom(domain, BaseRing(QR)));
    }

    CoCoA_ERROR(ERR::CanonicalHomFail, "CanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


  RingHom TmpChainCanonicalHom(const ring& domain, const ring& codomain)
  {
    //    std::clog << " TmpChainCanonicalHom ";
    try { return CanonicalHom(domain, codomain); }
    catch (const CoCoA::ErrorInfo& err) {if (err!=ERR::CanonicalHomFail) throw;}
    if (IsFractionField(codomain))
    {
      //      std::clog << " FrF ";
      return EmbeddingHom(codomain) (TmpChainCanonicalHom(domain, BaseRing(codomain)));
    }
    if (IsPolyRing(codomain))
    {
      //      std::clog << " P ";
      return CoeffEmbeddingHom(codomain) (TmpChainCanonicalHom(domain, CoeffRing(codomain)));
    }
    if (IsQuotientRing(codomain))
    {
      //      std::clog << " QR ";
      const QuotientRing QR = codomain;
      return QuotientingHom(QR) (TmpChainCanonicalHom(domain, BaseRing(QR)));
    }
    CoCoA_ERROR(ERR::CanonicalHomFail, "TmpChainCanonicalHom(R1,R2)");
    return IdentityHom(codomain); // Never executed; just to keep the compiler quiet.
  }


} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/CanonicalHom.C,v 1.10 2014/07/08 13:14:40 abbott Exp $
// $Log: CanonicalHom.C,v $
// Revision 1.10  2014/07/08 13:14:40  abbott
// Summary: Removed AsQuotientRing; added new defn of BaseRing
// Author: JAA
//
// Revision 1.9  2014/07/08 08:33:18  abbott
// Summary: Removed AsFractionField
// Author: JAA
//
// Revision 1.8  2014/07/07 12:12:17  abbott
// Summary: Removed AsPolyRing
// Author: JAA
//
// Revision 1.7  2012/02/10 10:26:40  bigatti
// -- changed RingZ.H, RingQ.H --> RingZZ.H, RingQQ.H
//
// Revision 1.6  2012/02/08 15:07:08  bigatti
// -- changed: Z,Q -> ZZ,QQ
//
// Revision 1.5  2011/02/18 12:56:08  bigatti
// -- added TmpChainCanonicalHom
//
// Revision 1.4  2009/07/24 14:21:03  abbott
// Cleaned up include directives, and added some fwd decls.
//
// Revision 1.3  2009/07/24 12:27:46  abbott
// Added some comments.
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.2  2007/03/07 11:18:01  bigatti
// -- fixed CanonicalHom goto flag
//
// Revision 1.1  2007/03/05 21:25:02  cocoa
// New CanonicalHom pseudo-ctor.
//
//
