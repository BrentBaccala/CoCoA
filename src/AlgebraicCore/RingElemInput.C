//   Copyright (c)  2014  Anna Bigatti

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


// Source code for input of polynomials and polynomial expressions

#include "CoCoA/RingElemInput.H"

#include "CoCoA/BigInt.H"
#include "CoCoA/error.H"
#include "CoCoA/ring.H"
#include "CoCoA/symbol.H"

#include <istream>
using std::istream;
//#include <iostream> // just for debugging
#include <sstream>
using std::istringstream;
#include <string>
using std::string;


namespace CoCoA
{

  char WhatsNext(std::istream& in)
  { // recognized characters
    in >> std::ws; // skip white spaces
    //    std::cout << " WhatsNext: testing " << char(in.peek()) << std::endl;
    if (in.eof()) return '\0';
    if (isdigit(in.peek())) return 'd'; // d for digit
    if (isalpha(in.peek())) return 'a'; // a for alpha
    if (in.peek()=='+') return '+';
    if (in.peek()=='-') return '-';
    if (in.peek()=='*') return '*';
    if (in.peek()=='/') return '/';
    if (in.peek()=='^') return '^';
    if (in.peek()=='(') return '(';
    if (in.peek()==')') return ')';
    if (in.peek()==';') return ';';
    CoCoA_ERROR("Illegal char \'"+string(1,char(in.peek()))+"\'","WhatsNext");
    return 'o';
  }


  void CoCoA_ERROR_NextChar(std::istream& in, const string& FuncName)
  {
    in.clear();
    CoCoA_ERROR("Unexpected \'"+ string(1,char(in.peek())) +"\'", FuncName);
  }


  RingElem ReadFactor(const ring& P, std::istream& in)
  {
    symbol s("temp");
    BigInt n;
    RingElem tmp(P);

    switch ( WhatsNext(in) )
    {
    case 'a':
      in >> s;  if (!in) CoCoA_ERROR_NextChar(in, "ReadFactor: symbol");
      tmp = RingElem(P, s);
      break;
    case 'd':
      in >> n;  if (!in) CoCoA_ERROR_NextChar(in, "ReadFactor: BigInt");
      tmp = RingElem(P, n);
      break;
    case '(':
      in.ignore(); // '('
      tmp = ReadExpr(P, in);
      if ( WhatsNext(in) != ')' )  CoCoA_ERROR_NextChar(in, "ReadFactor: ()");
      in.ignore(); // ')'
      break;
    default: CoCoA_ERROR_NextChar(in, "ReadFactor");
    }
    if ( WhatsNext(in) != '^' )  return tmp;
    in.ignore(); // '^'
    in >> n;  if (!in) CoCoA_ERROR_NextChar(in, "ReadFactor: exponent");
    return power(tmp, n);
  }


  RingElem ReadProduct(const ring& P, std::istream& in)
  {
    char w = '*';
    RingElem resProd(one(P));

    while (true)
    {
      if ( w == '*' )  resProd *= ReadFactor(P, in);
      else             resProd /= ReadFactor(P, in); // ( w == '/' )
      if ( WhatsNext(in) != '*' &&  WhatsNext(in) != '/' )  break;
      // uncomment the following to prevent a/b*c, a/b/c
      // if ( w == '/' ) 
      //   CoCoA_ERROR_NextChar(in, "ReadProduct: ambiguous after \'/\'");
      in >> w; // '*' or '/'
    }
    return resProd;
  }


  RingElem ReadExpr(const ring& P, std::istream& in)
  {
    char w = WhatsNext(in);
    RingElem f(P);

    while (true)
    {
      if ( w == '-' || w == '+' )  in.ignore(); // '+/-'
      if ( w == '-' )  f -= ReadProduct(P,in);  else  f += ReadProduct(P,in);
      w = WhatsNext(in);
      if ( (w == ';') || (w == ')') || (w == '\0') )  break;
      if ( (w != '-') && (w != '+') )  CoCoA_ERROR_NextChar(in, "ReadExpr");
    }
    return f;
  }


  RingElem ReadExprSemicolon(const ring& P, std::istream& in)
  {
    RingElem f = ReadExpr(P, in);
    if ( WhatsNext(in) != ';' )  CoCoA_ERROR_NextChar(in, "ReadExprSemicolon");
    in.ignore(); // ';'
    return f;
  }


  //-------- input from string --------------------
  RingElem ReadExprSemicolon(const ring& P, const std::string& s)
  {
    istringstream is(s);
    return ReadExprSemicolon(P, is);
  }


  RingElem ReadExpr(const ring& P, const std::string& s)
  {
    istringstream is(s);
    return ReadExpr(P, is);
  }


} // end of namespace CoCoA

// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/RingElemInput.C,v 1.3 2014/03/21 15:57:19 bigatti Exp $
// $Log: RingElemInput.C,v $
// Revision 1.3  2014/03/21 15:57:19  bigatti
// -- Ring is now first argument of ReadExpr
//
// Revision 1.2  2014/01/30 16:21:53  bigatti
// -- removed restriction about 1/3*x
//
// Revision 1.1  2014/01/30 15:15:33  bigatti
// -- first import
//
