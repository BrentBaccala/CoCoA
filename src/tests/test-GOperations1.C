//   Copyright (c)  2007-2012  John Abbott, Anna M. Bigatti

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/CanonicalHom.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/FreeModule.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/TmpGOperations.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/ModuleOrdering.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingElemInput.H"
#include "CoCoA/RingFp.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/VectorOperations.H"
#include "CoCoA/submodule.H"
#include "CoCoA/symbol.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for ideal/module operations (SAME EXAMPLES AS cocoa5.ts)
// functions: [GBasis], NF, IsElem, IsContained
// environments: DMP (Fp, Q), DMPI (Q), DMPII
//----------------------------------------------------------------------

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)


//[Id = "cocoa5-01", Descr = "ReducedGBasis5x Module posto"];
void test1()
{
  //Use Q[x,y,z],PosTo;
  SparsePolyRing P(NewPolyRing(RingQQ(),symbols("x","y","z")));
  RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2);
  FreeModule FM = NewFreeModule(P, 2, WDegPosnOrd);
  const vector<ModuleElem>& e = gens(FM);
  vector<ModuleElem> g;
  g.push_back(x*e[0] + z*e[1]);
  g.push_back(z*e[0] + y*e[1]);
  FGModule M = submodule(FM, g);
  cout << "TidyGens(M) = " << TidyGens(M) << endl;
  // [Vector(z, y), Vector(x, z), Vector(0, xy - z^2)];
}


//[Id = "cocoa5-02", Descr = "ReducedGBasis5x Module ToPos"];
void test2()
{
  //Use Q[x,y,z],ToPos;
  SparsePolyRing P(NewPolyRing(RingQQ(),symbols("x","y","z")));
  RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2);
  FreeModule FM = NewFreeModule(P, 2, OrdPosn);
  const vector<ModuleElem>& e = gens(FM);
  vector<ModuleElem> g;
  g.push_back(x*e[0] + z*e[1]);
  g.push_back(z*e[0] + y*e[1]);
  FGModule M = submodule(FM, g);
  cout << "TidyGens(M) = " << TidyGens(M) << endl;
  // [Vector(z, y), Vector(x, z)];
}


//Test := Record[Id = "cocoa5-03", Descr = "GBasis5x param"];
void test3()
{
  //Use Z/(32003)[a, x,y], Ord(M);
  PolyRing Fpa = NewPolyRing(NewRingFp(32003), symbols("a"));
  ring K = NewFractionField(Fpa);
  PolyRing P(NewPolyRing(K,2));
  RingElem x = indet(P,0),  y = indet(P,1);
  RingElem a = RingElem(P, symbol("a"));
  //    I := Ideal((a-1)*x+(a^2+a)*y, (a+1)*x + y);
  ideal I = ideal((a-1)*x+(power(a,2)+a)*y, (a+1)*x + y);
  vector<RingElem> GB = TidyGens(I);
  TEST_ASSERT("cocoa5-3" && GB[0]==x && GB[1]==y);
}

// Test := Record[Id := "cocoa5-04", Descr := "GBasis5x Module shifts"];

// Test.Input :="
// Use QQ[x,y,z], Weights([1,2,4]);
// M := Module([x^2-y,1],[x^4-z,y^2]);
// Info5 := Record[];
// Info5.ModuleShifts := Mat([[0,2]]);
// X := GBasis5x(M, Info5);
// X = [ Vector(x^2 - y, 1), Vector(y^2 - z, y^2 - x^2 - y)];
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-05", Descr := "ReducedGBasis5 Module OrdMat"];

// Test.Input :="
// OrdMat := Mat([[1,1,1],[2,1,1],[1,1,0]]); // Ring Grading (first 2 Rows) 
//                                           // Plus order (last row)
// Use ZZ/(101)[x,y,z], Ord(OrdMat), ToPos;
// M := Module([y-x,0,0], [x,0,z], [0,y^2-z^2,0]);
// X := ReducedGBasis5(Module(Gens(M)));
// X= [Vector(x,0,z),Vector(y,0,z),Vector(0, 0, x*z-y*z),Vector(0, y^2-z^2, 0)];
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-06", Descr := "ReducedGBasis5x Module shifts gradingdim2"];

// Test.Input :="
// OrdMat := Mat([[1,1,1],[2,1,1],[1,1,0]]); // Ring Grading (first 2 Rows) 
//                                           // Plus order (last row)
// Use ZZ/(101)[x,y,z], Ord(OrdMat), ToPos;
// M := Module([y-x,0,0], [x,0,z], [0,y^2-z^2,0]);
// Info5 := Record[];
// Info5.OrdMat := OrdMat;
// Info5.GradingDim := 2;
// Info5.ModuleShifts := Mat([[3,1,2],[2,2,5]]); // GrDim rows!!
// X := ReducedGBasis5x(M, Info5);
// X=[Vector(0, y^2-z^2, 0),Vector(x, 0, z),Vector(y, 0, z),Vector(0, 0, x*z - y*z)];
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-07", Descr := "Elim5"];

// Test.Input :="
// Use QQ[x,y,z,w[3..5]], Weights([7, 4, 3, 1, 1, 1]);
// // Use ZZ/(7)[x,y,z,w[3..5]], Weights([7, 4, 3, 1, 1, 1]); // exbug

// I := Ideal(
// 	   x - 7413431*w[4]^7 - 9162341*w[3]*w[4]*w[5]^5,
// 	   y - 6521443*w[4]^4 - 2312257*w[3]^2*w[4]*w[5],
// 	   z - 5329421*w[4]^3 - 2122414*w[3]*w[5]^2
// 	   );

// E  := Elim([w[3],w[4]], I);
// E5 := Elim5([w[3],w[4]], I);
// E=E5;
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-08", Descr := "Intersection5"];

// Test.Input :="
// Use QQ[x,y,z], Weights(1,2,1);
// I := Ideal(x*y, z^2);
// J := Ideal(y*z, x-z);
// I5 := Intersection5(I, J);
// I4 := Intersection(I, J);
// I4 = I5;
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-09", Descr := "Intersection5 ModuleShifts"];

// Test.Input :="
// Use QQ[x,y,z], Weights(1,2,1);
// M := Module([x^2*y, 0], [x*z^2, 0], [-x*z + x^2, -z + x]);
// N := Module([x*z^2, 0], [x*y*z, y*z], [-x*z + x^2, -z + x]);
// I5 := Intersection5x(M, N, Record[ModuleShifts := [[1,2]]]);
// I4 := Intersection(M, N);
// I4 = I5;
// ";    CoCoAServerRegister(Test, 1 /* times True */);

// -------------------------------
// -- TEST
// Test := Record[Id := "cocoa5-10", Descr := "Intersection5 Module"];

// Test.Input :="
// Use ZZ/(101)[x,y,z,t],PosTo;
// M:=Module([(x+y+t)^3,y^3],[(x-y-z)^3,z^3],[x^2-y^2,x^2-z^2],[x*y*z+y*z*t+z*t*x,x^3]);
// N:=Module([x^2+x*y,y^2]);

// I4 := Intersection(M,N);// Long
// I5 := Intersection5(M,N);// Long translation
// I4 = I5;
// ";    CoCoAServerRegister(Test, 1 /* times True */);


//[Id = "cocoa5-11", Descr = "Intersection5x 2 param"];
void test11()
{
  SparsePolyRing R = NewPolyRing(NewZZmod(32003), symbols("a"));
  FractionField K = NewFractionField(R);
  SparsePolyRing P(NewPolyRing(K,2));
  RingElem a = RingElem(P, symbol("a"));
  RingElem x = indet(P,0);
  RingElem y = indet(P,1);
// II := Ideal(x-y);
// I := (a-1) * x * II;
// J := (a+1) * y * II;
  ideal I = ideal((a-1) * x * (x-y));
  ideal J = ideal((a+1) * y * (x-y));
// X := Intersection5x(I,J, Record(NumParams=1));
  ideal X = intersect(I, J);
// X = x*y*II;
  TEST_ASSERT("cocoa5-11" && X == ideal(x*y*(x-y)));
}



//[Id = "cocoa5-26",Descr = "Homogenized5x "];
void test26()
{
// M1:=Mat([[2, 1, 1,0],
//          [1, 2, 0,1],
//          [0, 1, 0,1],
//          [1,0,0,0]]);
// Use Q[x,y,a,b],Ord(M1),PosTo;
// Info5 := Record();
// Info5.GradingDim := 2;
// Info5.OrdMat :=M1;
// I := Ideal(x+1,y^2+x);
// X := Homogenized5x([a,b],I,Info5);
// X = Ideal(a^2b + x, xb^3 + y^2, x^2b^2 - y^2a^2, y^2a^4 + x^3b);
}


//[Id = "cocoa5-28",Descr = "Homogenized5 "];
void test28()
{
// Use Q[x,y,z,h];
// I := Ideal(x^2y+xy+1, xy^2+y^2+1);
// X := Homogenized5([h],I);
// X = Ideal(x - y, y^3 + y^2h + h^3);
  SparsePolyRing P = NewPolyRing(RingQQ(), symbols("x","y","z","h"));
  RingElem x = RingElem(P,2);
  RingElem y = RingElem(P,3);
  RingElem z = RingElem(P,5);
  RingElem h = RingElem(P,7);
  RingElem y2 = power(y,2);
  ideal I(power(x,2)*y+x*y+1, x*y2+y2+1);
  ideal J(homog(I, h));
  TEST_ASSERT("cocoa5-28" && J == ideal(x - y, power(y,3) + y2*h + power(h,3)));
}


//[Id = "cocoa5-30", Descr = "Intersection5x multi-param"];
void test30()
{
// II := Ideal(x-y);
// I := (a+b-1) * x * II;
// J := (a-b^2+1) * y * II;
// X := Intersection5x(I,J, Record(NumParams=2));
// X=x*y*II;
  SparsePolyRing R(NewPolyRing(NewZZmod(32003),symbols("a","b")));
  SparsePolyRing P(NewPolyRing(NewFractionField(R), symbols("x","y")));

  ideal II = ideal(ReadExpr(P, "x-y"));
  ideal I = ideal(ReadExpr(P, "x * (a+b-1)")) * II;
  ideal J = ideal(ReadExpr(P, "y * (a-b^2+1)")) * II;
  ideal X = intersect(I,J);
  TEST_ASSERT("cocoa5-30" && X == ideal(ReadExpr(P, "x*y"))*II);
}


void program()
{
  GlobalManager CoCoAFoundations;
  
  test1();
  test2();
  test3();
  test11();
  test30();
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
