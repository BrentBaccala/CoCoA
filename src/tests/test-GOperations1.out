TidyGens(M) = [[z, y],  [x, z],  [0, x*y -z^2]]
TidyGens(M) = [[z, y],  [x, z]]
